InteralsVisibleToCompletionProvider: Fix insertion of wrong assembly name if name has dots

MaStr11 opened the pull request:
Fixes #29446.

MaStr11 commented on src/Features/Core/Portable/Completion/Providers/AbstractInternalsVisibleToCompletionProvider.cs:
```diff
@@ -239,10 +239,32 @@ public override async Task<CompletionChange> GetChangeAsync(Document document, C
+                {
+                    span = new TextSpan(
+                        start: token.Span.Start + positionOfValueIntText,
+                        length: token.ValueText.Length);
```
This approach is horrible, but I couldn't figure out how to make sure the completion item spans the whole string while preserving any existing double quotes. @CyrusNajmabadi you probably know a better way how to fix this.

CyrusNajmabadi replied:
There are two approaches you can take here.  I'm fine with either:

1. Walk backwards until you hit the quote character.   Not great, but workable.  Would be incorrect if there was a `\"` in the string.
2. you have the start of the string.  and you can find out if it's regular or varbatim.  if it's regular, then the position you want is `start + "\"".Length`.  If it's verbatim you want `start + "@\"".Length`.

I'd prefer approach #2.

CyrusNajmabadi commented on src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests_InternalsVisibleTo.vb:
```diff
@@ -598,5 +598,85 @@ using System.Reflection;
+                Await state.WaitForAsynchronousOperationsAsync()
+                state.AssertMatchesTextStartingAtLine(1, "[assembly: System.Runtime.CompilerServices.InternalsVisibleTo(""Dotted1.Dotted2.Assembly.Dotted3"")]")
+            End Using
+        End Function
```
can we have test using a verbatim string literal? i.e. `@""`?

ivanbasov replied:
Please add the test

---
In reply to: [212065955](https://github.com/dotnet/roslyn/pull/29447#discussion_r212065955) [](ancestors = 212065955)

MaStr11 replied:
Done `CodeCompletionReplacesExisitingAssemblyNameWithDots_Verbatim_1` and `CodeCompletionReplacesExisitingAssemblyNameWithDots_Verbatim_2`

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractInternalsVisibleToCompletionProvider.cs:
```diff
@@ -239,10 +239,32 @@ public override async Task<CompletionChange> GetChangeAsync(Document document, C
+            if (syntaxFactsService.IsStringLiteral(token) && !string.IsNullOrEmpty(token.ValueText))
+            {
+                span = token.Span;
+                var positionOfValueIntText = token.Text.IndexOf(token.ValueText);
```
nope.  do not do this.  it will not work in any sort of interesting cases (i.e. an escaped character).  In fact, please add a test that included an escaped character right now :)

ivanbasov replied:
@MaStr11, please consider suggestions by @cyrus.najmabadi@gmail.com 

---
In reply to: [212066374](https://github.com/dotnet/roslyn/pull/29447#discussion_r212066374) [](ancestors = 212066374)

MaStr11 commented:
@CyrusNajmabadi The following changes were made since your last review:

### Moved the fix of the span to an earlier stage 

Before: The span was fixed for the TextChange describing the commit. Now: The fix is done on the completion item. I think that is the correct way because now the pre-selection of the best fitting assembly should also work. I haven't added any tests for the preselection though. This is the doc that makes me think that it is better suited:
```CS
        /// <summary>
        /// The span of the document the completion list corresponds to.  It will be set initially to
        /// the result of <see cref="CompletionService.GetDefaultCompletionListSpan"/>, but it can
        /// be overwritten during <see cref="CompletionService.GetCompletionsAsync"/>.  The purpose
        /// of the span is to:
        ///     1. Signify where the completions should be presented.
        ///     2. Designate any existing text in the document that should be used for filtering.
        ///     3. Specify, by default, what portion of the text should be replaced when a completion 
        ///        item is committed.
        /// </summary>
        public TextSpan CompletionListSpan { get; set; }
```
### Changed the logic on how to fix the TextSpan.

I follow your advice and also added some tests.

Note: The `CodeCompletionReplacesExisitingAssemblyNameWithDots_OpenEnded` test fails as of now. What needs to be done to make it pass is to expand the TextSpan until a new line starts. But before I implement this I wanted to make sure I'm on the right track in general.

My concerns with the current approach are:

* It depends on double quotes as indicators of text start and text end. This seem to be working for C# and VB for now but it feels like it is something language specific.
* It feels like the logic of how to get the span of the string literal content should be placed somewhere else as this is likely something that can be usefull for others too (maybe it is already solved somewhere else within Roslyn?).

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractInternalsVisibleToCompletionProvider.cs:
```diff
@@ -148,6 +148,11 @@ private async Task AddAssemblyCompletionItemsAsync(CompletionContext context, Ca
+
+            if (context.Items.Count > 0)
+            {
+                context.CompletionListSpan = await GetTextChangeSpanAsync(context.Document, context.CompletionListSpan, cancellationToken);
```
needs ConfigureAwait. also wrap parameter line to make more visible.

ivanbasov replied:
yes, please add ConfigureAwait

---
In reply to: [212387444](https://github.com/dotnet/roslyn/pull/29447#discussion_r212387444) [](ancestors = 212387444)

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractInternalsVisibleToCompletionProvider.cs:
```diff
@@ -243,6 +248,32 @@ public override async Task<CompletionChange> GetChangeAsync(Document document, C
+                var contentStartsAt = token.Text.IndexOf('"');
+                contentStartsAt = contentStartsAt >= 0
+                    ? contentStartsAt + 1
+                    : contentStartsAt;
```
nit.  if contentStartsAt was less than 0, then i wouldn't do any work.

MaStr11 replied:
I check if the token is a (Verbatim)StringLiteral before. I think IndexOf will never return -1. Should I simplify that to
`var contentStartsAt = token.Text.IndexOf('"')+1;`?

CyrusNajmabadi replied:
> I think IndexOf will never return -1. 

I don't trust things that much.  I would assume the worst and have the code check for that.

Note: i don't think you need any of this if we use the GetWordSpan approach suggested above.  That would be my preference on how we code up this solution.

ivanbasov commented:
@jinujoseph , the code involves completion tests. I would suggest the code to be sent against 'master_vs_deps' rather than 'master'. What do you think?

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractInternalsVisibleToCompletionProvider.cs:
```diff
@@ -243,6 +248,32 @@ public override async Task<CompletionChange> GetChangeAsync(Document document, C
+                    ? token.Span.End
+                    : token.Text[token.Span.Length - 1] == '"'
+                        ? token.Span.End - 1
+                        : startSpan.End; // The string is open ended. We keep the right end as is.
```
do you have tests for this case?

MaStr11 replied:
Yes. A few. Including one that's failing for now (See also my comment in the thread). I think the best would be to extend the span to the right until the end of line or some whitespace. What do you think.

CyrusNajmabadi replied:
Yes.  that woudl fit how the normal "compute the text span" logic is intended to work.


In fact, to make your life easier.  Just call: CommonCompletionUtilities.GetWordSpan.  However, for the isWordCharacter helper pass: `ch => SyntaxFacts.IsIdentifierPartCharacter(ch) || ch == '.'`

MaStr11 replied:
This doesn't work if you there are escape sequences before the cursor.

CyrusNajmabadi replied:
Do we ever expect there to be escape characters in an assembly string like this?  I highly doubt it.  And, as such, i think it's fine to have the simple code that works for the 99.99% case :)

CyrusNajmabadi replied:
Note: if you do want to support escapes, why not just add \ into the list of acceptable characters to consume?

jinujoseph commented:
@ivanbasov sure your can target deps for this

MaStr11 commented:
> re-target the PR to master-vs-deps

I did. Unfortunately, this triggered a review request for some unrelated teams.

MaStr11 commented on src/Features/Core/Portable/Completion/Providers/AbstractInternalsVisibleToCompletionProvider.cs:
```diff
@@ -243,6 +248,49 @@ public override async Task<CompletionChange> GetChangeAsync(Document document, C
+                    end++;
+                }
+
+                return end;
```
@CyrusNajmabadi you proposed to do:

>In fact, to make your life easier. Just call: CommonCompletionUtilities.GetWordSpan. However, for the isWordCharacter helper pass: ch => SyntaxFacts.IsIdentifierPartCharacter(ch) || ch == '.'

`CommonCompletionUtilities.GetWordSpan` moves the end further only if the char before the cursor is a word character. Therefore I decided to make a copy of the relevant code. I do also think that we should not rely on `IsIdentifierPartCharacter` here, because we expect an assembly name with an optional `, PublicKey=hash` appendix. So a proper character list would be 
* all chars allowed in an assembly name
* comma, space, and equals sign

CyrusNajmabadi replied:
> moves the end further only if the char before the cursor is a word character. Therefore I decided to make a copy of the relevant code.

I'd much rather us tweak GetWordSpan so it can handle that case.  something like `bool alwaysExtendEndSpan = false`.

CyrusNajmabadi replied:
> So a proper character list would be...

Sure, that seems reasonable to me.  :)

MaStr11 replied:
Done in 30340f4.

MaStr11 replied:
I only use it to find the end. To find the start I do start at the begin of the token.Span to avoid problems with escape sequences.

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractInternalsVisibleToCompletionProvider.cs:
```diff
@@ -243,6 +248,40 @@ public override async Task<CompletionChange> GetChangeAsync(Document document, C
+                    ? token.Span.End
+                    : token.Text[token.Span.Length - 1] == '"'
+                        ? token.Span.End - 1
+                        : await GetEndOfOpenEndedStringLiteral(startSpan.End).ConfigureAwait(false); // The string is open ended. We keep the right end as is.
```
i would just always do this.  It will work properly for open and close ended strings just fine.

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractInternalsVisibleToCompletionProvider.cs:
```diff
@@ -243,6 +248,40 @@ public override async Task<CompletionChange> GetChangeAsync(Document document, C
+            async Task<int> GetEndOfOpenEndedStringLiteral(int startPosition)
+            {
+                var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                var span = CommonCompletionUtilities.GetWordSpan(text, startPosition, ch => false, ch => !char.IsWhiteSpace(ch), alwaysExtendEndSpan: true);
```
basically, you could consume until you hit a quote or newline char.  that would consume the right amount in both directions.

MaStr11 replied:
Except escaped double quotes. That's why I used the approach with the StringLiteralToken. But I'm with you: Escaped double quotes are highly unlikely in this context (and maybe even illegal).

I will give it a try later as it should indeed be much simpler code.

CyrusNajmabadi replied:
👍 .  I'm 100% ok with this not doing a perfect job in such esoteric cases.

MaStr11 commented:
@CyrusNajmabadi I did the following since we last talked:

1. Simplified the calculation of the span e7a852a
1. Skip the former passing test with double quote escape sequences 
a045e48 (How should we document that we don't support double quotes: Skip the test or adopt the test to reflect the current behavior)
1. Added a test case with a public key being present 3444638
1. Added (partial) support for cases when the public key spans multiple lines c29a7de (parts of the old key are preserved).

CyrusNajmabadi commented:
> How should we document that we don't support double quotes: Skip the test or adopt the test to reflect the current behavior)

adopt the test to reflect hte current behavior.  add a comment in the test indicating that this isn't perfect, but is acceptable given the simplicity it allows the code to be.

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractInternalsVisibleToCompletionProvider.cs:
```diff
@@ -227,6 +238,26 @@ private async Task<string> GetAssemblyNameFromInternalsVisibleToAttributeAsync(D
+                var text = root.GetText();
+                
+                // Expand selection in both directions until a double quote or any line break character is reached
+                bool IsWordCharacter(char ch) => !(ch == '"' || (char.IsControl(ch) || char.IsSeparator(ch) && !char.IsWhiteSpace(ch)));
```
1. this is super hard to read.
2. it doesn't match your comment.
3. why not actually just check for double-quote/line-break?  why the need for Control/Separator/Whitespace??

MaStr11 replied:
I wanted to check for the unicode categories LineSeparator and ParagraphSeparator but.netstandard 1.3 doesn't provide a unicode category method. So I used the IsSeparator method and fixed what was missing and what was too much. I will change it to explicit call out the line break characters instead.

I was also thinking about using your IVirtualCharService to get the bounds of the text inside the StringLiteralToken. Then I only need to make sure that the span stays on the same line (It seems at least in VB the completion fails if the CompletionSpan spans more then one line). What do you think about that approach?

CyrusNajmabadi replied:
I think it's overkill :).  We know what 99.9999% of all IntervalsVisibleTo strings will look like.  It seems like it should just be sufficient to eat forward/backward until you hit one of:

1. quote
2. end-of-line
3. end-of-file

MaStr11 commented on src/Features/Core/Portable/Completion/Providers/AbstractInternalsVisibleToCompletionProvider.cs:
```diff
@@ -249,7 +249,7 @@ private static async Task<TextSpan> GetTextChangeSpanAsync(Document document, Te
                 
                 // Expand selection in both directions until a double quote or any line break character is reached
-                bool IsWordCharacter(char ch) => !(ch == '"' || ch == '\u000A' || ch == '\u000C' || ch == '\u000D' || ch == '\u2028' || ch == '\u2029');
+                bool IsWordCharacter(char ch) => !(ch == '"' || ch== '\u0003' ||  ch == '\u000A' || ch == '\u000C' || ch == '\u000D' || ch == '\u2028' || ch == '\u2029');
```
TextUtilities is internal and can't be used here. I copied the character list from there for now.

CyrusNajmabadi replied:
Add it as a linked file into the workspace.  we have ample precednece for these sorts of basic helpers :)  (i thought we'd already done that here).  Tnx! :)

CyrusNajmabadi replied:
it can go under: Workspace/Utilities/CompilerUtilities .

MaStr11 replied:
I just added it to Shared\Utilities. Is this also ok?

CyrusNajmabadi replied:
Note: you don't need to handle `\u0003` explicitly.  It was my bad saying "end of file".  What i meant was: you are at the end of the file.  Roslyn doesn't actually represent this with an actual \u0003 character.

CyrusNajmabadi replied:
shared\utilities works for me :)

CyrusNajmabadi replied:
Note: to be clear, i don't think you have to add any 'end of file' checks.   i think that is already handled for you by the GetWordSpan code.  However, i do think we need tests here.

CyrusNajmabadi commented on src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests_InternalsVisibleTo.vb:
```diff
@@ -639,5 +639,220 @@ using System.Reflection;
+                state.AssertMatchesTextStartingAtLine(1, "[assembly: System.Runtime.CompilerServices.InternalsVisibleTo(""Dotted1.Dotted2.Assembly.Dotted3"" +")
+                state.AssertMatchesTextStartingAtLine(2, "                                                              ""0000000602000000240000525341310004000"" +")
+            End Using
+        End Function
```
do we have a test with multiple forms of unterminated strings?  i think we need a test where it's unterminated, and has a newline, and one where it is unterminated and the end of the file is on the same line.

MaStr11 replied:
I added some explicit tests anyway.

ivanbasov commented:
Thank you!

---
In reply to: [415810725](https://github.com/dotnet/roslyn/pull/29447#issuecomment-415810725) [](ancestors = 415810725)

ivanbasov commented:
Thank you very much, @MaStr11 and @CyrusNajmabadi for your help!

MaStr11 commented:
Thank you @CyrusNajmabadi for your guidance!
